oak-core/src/main/java/org/apache/jackrabbit/oak/security/user/ImpersonationImpl.java

```diff
@@ -173,13 +182,49 @@ private void updateImpersonatorNames(@NotNull Tree userTree, @NotNull Set<String
     private boolean isAdmin(@NotNull Principal principal) {
         if (principal instanceof AdminPrincipal) {
             return true;
-        } else if (GroupPrincipals.isGroup(principal)) {
+        }
+        return Utils.isAdmin(principal, user.getUserManager());
+    }
+
+    private boolean isImpersonator(@NotNull Set<String> principalNames) {
+        Set<String> impersonatorGroups = Set.of(user.getUserManager().getConfig().getConfigValue(
+                PARAM_IMPERSONATOR_GROUPS_ID,
+                new String[]{}));
+
+        if (impersonatorGroups.isEmpty()) {
             return false;
-        } else {
-            return Utils.canImpersonateAllUsers(principal, user.getUserManager());
         }
+        return principalNames.stream()
+                .anyMatch(impersonatorGroups::contains);
     }
 
+    public boolean isImpersonator(@NotNull Authorizable authorizable) {
+        Set<String> impersonatorGroups = Set.of(user.getUserManager().getConfig().getConfigValue(
+                PARAM_IMPERSONATOR_GROUPS_ID,
+                new String[]{}));
+
+        if (impersonatorGroups.isEmpty()) {
+            return false;
+        }
+
+        try {
+            Iterator<Group> it =  authorizable.memberOf();
```

Comment:
thinking about this a bit longer, i also wonder, why you loop over the membership of the user, which might contain something between multiple to many groups.

so for the query part of the story, why not doing something along the following lines:

```
Principal userPrincipal = authorizable.getPrincipal();
for (String principalName : configuredImpersonators) {
     Principal principal = principalManager.getPrincipal(principalName);
     if (principal != null) {
          if (GroupPrincipals.isGroup(principal)) {
               if (((GroupPrincipal) principal).isMember(userPrincipal)) {
                    return true;
               }
          } else if (principalName.equals(userPrincipal.getName()) {
               return true;
          }
     }
}
return false;
```

the reasons:
- i would expect the configuration to only contain a very limited number of entries (if we fear that it might become big we can even limit the cardinality of the configuration0
- instead of looping over all groups (and for most cases not finding a matching group) we check directly if the user-principal is member of any of the group principals (lets say there are 1 or 2 entries in the config only)
- for non-group configuration entries we just check for a matching name.
- for configuration entries that link to non-existing prinicipals -> we ignore it
- for default local group princpals stored in the user manager that might also be the better way due to the way group membership is stored.

one more thing: this way we are also able to cover the edge case where a configured principal name that is not stored in the repository as a regular group but would still end up in the subject of a given user (yes, oak allows to plug custom sources for principals. i can show it to you if you want).

Reply:
I like the edge case that your proposal is also covering, might be something wanted but nobody thought of. 
I believe the number of groups somebody is member of is also small most of the times,  so it comes up to what is more expensive: 
a)"memberOf" + looping through list of groups a principal is member of
b) getting groups from the config and comparing them with the principal
I believe b) is faster.

_I have a question_: shouldn't we loop through the "principals of the authorizable" when evaluating membership? By this I mean getting the subject for the current `authorizable.getPrincipal()` and then get all the principals associated to that subject ? If YES, I didn't know a way of getting the subject, this is why I chose looping through the groups of a principal.